File: Classic/quickSort.cpp
/* ©ЛкыеепР */
#include <iostream>

using namespace std;

void quickSort(int* a, int left, int right) {
	if (left < right) {
		int key = a[left];
		int low = left;
		int high = right;
		
		while (low < high) {
			while (low < high && a[high] > key) {
				high--;
			}	
			a[low] = a[high];
			while (low < high && a[low] < key) {
				low++;
			}
			a[high] = a[low];
		}
		
		a[low] = key;
		
		quickSort(a, left, low - 1);
		quickSort(a, low + 1, right);
	}
}

int main() {
	int a[] = {2, 4, 9, 3, 6, 7, 1, 5};
	
	quickSort(a, 0, 7);
	
	for (int i = 0; i < sizeof(a); i++) {
		cout << a[i] << " ";
	}
	
	return 0;
}

File: basis/sort/quickSort.py
import sys


def partition(arr, left, right):
    i = left
    j = right
    k = arr[left]
    while i < j:
        while i < j and arr[j] > k:
            j -= 1
        if i < j:
            arr[i] = arr[j]
            i += 1
        while i < j and arr[i] < k:
            i += 1
        if i < j:
            arr[j] = arr[i]
            j -= 1
    arr[i] = k
    return i


def quickSort(arr, left, right):
    if left > right:
        return
    k = partition(arr, left, right)
    quickSort(arr, left, k - 1)
    quickSort(arr, k + 1, right)
    return arr


for line in sys.stdin:
    arr = list(map(int, line.split()))
    print(quickSort(arr, 0, len(arr) - 1))
